lectures/zksnark.tex

```diff
@@ -0,0 +1,408 @@
+\end{definition}
+
+\begin{example}
+    Let's get back to our example from the previous chapter about the arithmetic circuits.
+    \[ r = x_1 \times x_2 \]
+    The constraint is:
+    \[(a_1w_1 + a_2w_2 + a_3w_3)(b_1w_1 + b_2w_2 + b_3w_3) = c_1w_1 + c_2w_2 + c_3w_3\]
+    Coefficients and witness vectors are:
+    $w = (r, x_1, x_2), a = (0, 1, 0), b = (0, 0, 1), c = (1, 0, 0)$. Therefore:
+    \[ (0w_1 + 1w_2 + 0w_3)(0w_1 + 0w_2 + 1w_3) = (1w_1 + 0w_2 + 0w_3) \]
+    \[ w_2 \times w_3 = w_1 \]
+    \[ x_1 \times x_2 = r \]
+\end{example}
+
+
+\begin{example}
+    And a more complex example. Remember that we want to verify each computational step.
+
+    \begin{lstlisting}[language=C,numbers=none]
+    if (x1) {
+        return x2 * x3
+    } else {
+        return x2 + x3
+    }
+    \end{lstlisting}
+
+    We know that it can be expressed as:
+    \[ r = x_1 \times (x_2 \times x_3) + (1 - x_1) \times (x_2 + x_3) \]
+
+    However, one important consideration was overlooked. If $x_1$ is neither $0$ nor $1$, it implies
+    that something else is being computed instead of the desired program. Since we need to add a
+    restriction for $x_1$: $x_1 \times (1 - x_1) = 0$, this effectively checks that $x_1$ is binary.
+
+    The next constraints can be build:
+    \begin{align*}
+        x_1 \times x_1 &= x_1 \quad \text{(binary check)} \tag{1} \\
+        x_2 \times x_3 &= mult \tag{2} \\
+        x_1 \times mult &= selectMult \tag{3} \\
+        (1 - x_1) \times (x_2 + x_3) &= r - selectMult \tag{4}
+    \end{align*}
+
+    For every constraint we need the coefficients vectors $a_i$, $b_i$, $c_i$, but all of them have
+    the same witness vector $w$.
+
+    The interesting thing is where to take a constant from. In our case, it's just 1 in the (4) 
+    constraint. The solution is straightforward: by placing 1 in the witness vector, we can then 
+    obtain any desired value by multiplying it by an appropriate coefficient which is $-1$ in our
+    case.
+
+    The witness vector:
+    \[ w = (1, r, x_1, x_2, x_3, mult, selectMult) \]
+    The coefficients vectors:
+    \begin{align*}
+        a_1 &= (0, 0, 1, 0, 0, 0, 0) & \quad b_1 &= (0, 0, 1, 0, 0, 0, 0) & \quad c_1 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_2 &= (0, 0, 0, 1, 0, 0, 0) & \quad b_2 &= (0, 0, 0, 0, 1, 0, 0) & \quad c_2 &= (0, 0, 0, 0, 0, 1, 0) \\
+        a_3 &= (0, 0, 1, 0, 0, 0, 0) & \quad b_3 &= (0, 0, 0, 0, 0, 1, 0) & \quad c_3 &= (0, 0, 0, 0, 0, 0, 1) \\
+        a_4 &= (1, 0, -1, 0, 0, 0, 0) & \quad b_4 &= (0, 0, 0, 1, 1, 0, 0) & \quad c_4 &= (0, 1, 0, 0, 0, 0, -1)
+    \end{align*}
+
+    Now, let's use specific values to compute an example:
+    \[ x_1 = 1, \quad x_2 = 3, \quad x_3 = 4 \]
+
+    We get:
+    \[ \text{mult} = x_2 \times x_3 = 3 \times 4 = 12 \]
+    \[ \text{selectMult} = x_1 \times \text{mult} = 1 \times 12 = 12 \]
+    \[ r = x_1 \times (x_2 \times x_3) + (1 - x_1) \times (x_2 + x_3) \]
+    \[ r = 1 \times (3 \times 4) + (1 - 1) \times (3 + 4) = 12 \]
```

Comment:
Maybe I missed that, but specify somewhere that you are using $\mathbb{R}$ (or some large finite field $\mathbb{F}_p$) for all the arithmetic